packages/flutter/lib/src/material/segmented_button.dart

```diff
@@ -0,0 +1,726 @@
+    final OutlinedBorder enabledBorder = (resolvedEnabledBorder ?? const RoundedRectangleBorder()).copyWith(side: enabledSide);
+    final OutlinedBorder disabledBorder = (resolvedDisabledBorder ?? const RoundedRectangleBorder()).copyWith(side: disabledSide);
+    final OutlinedBorder? borderShape = resolvedEnabledBorder?.copyWith(side: BorderSide.none);
+
+    final List<Widget> buttons = segments.map(buttonFor).toList();
+
+    return Material(
+      shape: borderShape,
+      elevation: style.elevation?.resolve(_enabled ? enabledState : disabledState) ?? 0.0,
+      shadowColor: style.shadowColor?.resolve(_enabled ? enabledState : disabledState) ?? Colors.transparent,
+      clipBehavior: Clip.antiAlias,
+      child: _SegmentedButtonRenderWidget<T>(
+          segments: segments,
+          enabledBorder: _enabled ? enabledBorder : disabledBorder,
+          disabledBorder: disabledBorder,
+          direction: direction,
+          children: buttons,
+      ),
+    );
+  }
+}
+class _SegmentedButtonRenderWidget<T> extends MultiChildRenderObjectWidget {
+  _SegmentedButtonRenderWidget({
+    super.key,
+    required this.segments,
+    required this.enabledBorder,
+    required this.disabledBorder,
+    required this.direction,
+    required super.children,
+  }) : assert(children.length == segments.length);
+
+  final List<ButtonSegment<T>> segments;
+  final OutlinedBorder enabledBorder;
+  final OutlinedBorder disabledBorder;
+  final TextDirection direction;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderSegmentedButton<T>(
+      segments: segments,
+      enabledBorder: enabledBorder,
+      disabledBorder: disabledBorder,
+      textDirection: direction,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderSegmentedButton<T> renderObject) {
+    renderObject
+      ..segments = segments
+      ..enabledBorder = enabledBorder
+      ..disabledBorder = disabledBorder
+      ..textDirection = direction;
+  }
+}
+
+class _SegmentedButtonContainerBoxParentData extends ContainerBoxParentData<RenderBox> {
+  RRect? surroundingRect;
+}
+
+typedef _NextChild = RenderBox? Function(RenderBox child);
+
+class _RenderSegmentedButton<T> extends RenderBox with
+     ContainerRenderObjectMixin<RenderBox, ContainerBoxParentData<RenderBox>>,
+     RenderBoxContainerDefaultsMixin<RenderBox, ContainerBoxParentData<RenderBox>> {
+  _RenderSegmentedButton({
+    required List<ButtonSegment<T>> segments,
+    required OutlinedBorder enabledBorder,
+    required OutlinedBorder disabledBorder,
+    required TextDirection textDirection,
+  }) : _segments = segments,
+       _enabledBorder = enabledBorder,
+       _disabledBorder = disabledBorder,
+       _textDirection = textDirection;
+
+  List<ButtonSegment<T>> get segments => _segments;
+  List<ButtonSegment<T>> _segments;
+  set segments(List<ButtonSegment<T>> value) {
+    if (_segments == value) {
```

Comment:
I wonder if we should use [listEquals](https://api.flutter.dev/flutter/foundation/listEquals.html) here, since build methods are likely to rebuild the list of segments each time the selection changes.